core/src/main/scala/org/apache/spark/internal/Logging.scala

```diff
@@ -230,19 +235,18 @@ private class SparkShellLoggingFilter extends Filter {
    */
   def decide(loggingEvent: LoggingEvent): Int = {
     if (Logging.sparkShellThresholdLevel == null) {
-      return Filter.NEUTRAL
-    }
-    val rootLevel = LogManager.getRootLogger().getLevel()
-    if (!loggingEvent.getLevel().eq(rootLevel)) {
-      return Filter.NEUTRAL
-    }
-    var logger = loggingEvent.getLogger()
-    while (logger.getParent() != null) {
-      if (logger.getLevel() != null) {
-        return Filter.NEUTRAL
+      Filter.NEUTRAL
+    } else if (loggingEvent.getLevel.isGreaterOrEqual(Logging.sparkShellThresholdLevel)) {
+      Filter.NEUTRAL
+    } else {
+      var logger = loggingEvent.getLogger()
+      while (logger.getParent() != null) {
+        if (logger.getLevel != null || logger.getAllAppenders.hasMoreElements) {
```

Comment:
So I tested your code and it seems to work... but I can't figure out why. This is basically the only change I'm seeing in the code that would affect the outcome (previous code didn't check the appenders).

Could you use the PR description to explain the fix itself? All you've done is explain the problem. That's already explained in the bug you filed.

Reply:
I'll update the description.
Not checking the appenders is one of the defect of the current implementation but not the most important point.
The most important point is that whether a message is logged and output to the REPL's console is based on the log level between a message, root logger and REPL's logger.

Reply:
Ok, i tested again and the culprit is the line that does `if (!loggingEvent.getLevel().eq(rootLevel))`. which is what your updated description says. So all's good.